File: tests/Aiko.Core.Tests/ClaudeSettingsEditorTests.cs
using Aiko.Core;

namespace Aiko.Core.Tests;

/// Writing to a file that belongs to another program.
///
/// Removal walks every Claude Code folder on the machine and rewrites what it finds, and until now
/// none of it had a test. These cover the paths that end badly: no backup, a backup already there,
/// a file the user changed after installing, and a folder Aiko cannot read.
public class ClaudeSettingsEditorTests
{
    private const string Folder = @"C:\Users\someone\.claude";
    private const string Bridge = @"C:\Users\someone\AppData\Local\Aiko\current\Aiko.Bridge.exe";
    private const string Moved = @"D:\Aiko\Aiko.Bridge.exe";

    private static string Settings => ClaudeSettingsEditor.PathIn(Folder);

    private static string Backup => ClaudeSettingsEditor.BackupPathIn(Folder);

    private static string Command(string path = Bridge) => BridgeCommand.For(path);

    [Fact]
    public void A_copy_is_made_before_the_first_change()
    {
        var original = """{ "statusLine": { "type": "command", "command": "mine.sh" } }""";
        var files = new FakeFiles().With(Settings, original);

        new ClaudeSettingsEditor(files, Windows).Add(Folder, Command());

        Assert.True(files.Has(Backup));
        Assert.Equal(original, files.Read(Backup));
    }

    [Fact]
    public void A_copy_that_already_exists_is_left_alone()
    {
        // A second copy would save our own edit and lose what the user had.
        var files = new FakeFiles()
            .With(Settings, """{ "statusLine": { "type": "command", "command": "new.sh" } }""")
            .With(Backup, """{ "statusLine": { "type": "command", "command": "original.sh" } }""");

        new ClaudeSettingsEditor(files, Windows).Add(Folder, Command());

        Assert.Contains("original.sh", files.Read(Backup));
    }

    [Fact]
    public void A_folder_with_no_settings_file_gets_one()
    {
        var files = new FakeFiles();

        var outcome = new ClaudeSettingsEditor(files, Windows).Add(Folder, Command());

        Assert.True(outcome.Changed);
        Assert.True(files.Has(Settings));
        // Nothing existed, so there was nothing worth copying.
        Assert.False(files.Has(Backup));
    }

    [Fact]
    public void The_file_is_written_through_a_temporary_one()
    {
        // A half written settings file breaks Claude Code, not just Aiko.
        var files = new FakeFiles();

        new ClaudeSettingsEditor(files, Windows).Add(Folder, Command());

        Assert.Contains(files.Writes, w => w.EndsWith(".aiko.tmp", StringComparison.Ordinal));
        Assert.DoesNotContain(files.Writes, w => w == Settings);
    }

    [Fact]
    public void Adding_our_line_twice_writes_nothing_the_second_time()
    {
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());
        files.Writes.Clear();

        var outcome = editor.Add(Folder, Command());

        Assert.False(outcome.Changed);
        Assert.Empty(files.Writes);
    }

    [Fact]
    public void Removal_puts_back_the_line_the_user_had()
    {
        var files = new FakeFiles()
            .With(Settings, """{ "statusLine": { "type": "command", "command": "mine.sh" } }""");
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        var outcome = editor.Remove(Folder);

        Assert.True(outcome.Changed);
        Assert.Contains("mine.sh", files.Read(Settings));
        Assert.DoesNotContain(SettingsJsonPatch.WrappedKey, files.Read(Settings));
    }

    [Fact]
    public void Removal_drops_the_copy_it_made()
    {
        var files = new FakeFiles()
            .With(Settings, """{ "statusLine": { "type": "command", "command": "mine.sh" } }""");
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        editor.Remove(Folder);

        // The copy was insurance while Aiko was in the file, and the readme promises there is
        // nothing left to clean up by hand.
        Assert.False(files.Has(Backup));
    }

    [Fact]
    public void Removal_keeps_what_the_user_changed_after_installing()
    {
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        // They edited the file themselves while Aiko was in it.
        var edited = System.Text.Json.Nodes.JsonNode.Parse(files.Read(Settings))!.AsObject();
        edited["model"] = "opus";
        files.With(Settings, edited.ToJsonString());

        editor.Remove(Folder);

        // Restoring the copy wholesale would have thrown this away.
        Assert.Contains("opus", files.Read(Settings));
    }

    [Fact]
    public void Removal_deletes_a_settings_file_Aiko_made_from_nothing()
    {
        // Windows Sandbox: no settings.json before Aiko, and "{}" left behind after it.
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        editor.Remove(Folder);

        Assert.False(files.Has(Settings));
    }

    [Fact]
    public void Removal_keeps_a_file_Aiko_made_once_something_else_was_written_to_it()
    {
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        var edited = System.Text.Json.Nodes.JsonNode.Parse(files.Read(Settings))!.AsObject();
        edited["model"] = "opus";
        files.With(Settings, edited.ToJsonString());

        editor.Remove(Folder);

        Assert.True(files.Has(Settings));
        Assert.Contains("opus", files.Read(Settings));
    }

    [Fact]
    public void Removal_keeps_an_empty_file_that_was_there_before_Aiko()
    {
        // It was the user's file, even if it said nothing. A copy was made of it, and that copy is
        // what tells the two cases apart.
        var files = new FakeFiles().With(Settings, "{}");
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        editor.Remove(Folder);

        Assert.True(files.Has(Settings));
    }

    [Fact]
    public void Removal_from_a_folder_Aiko_never_touched_changes_nothing()
    {
        var theirs = """{ "model": "sonnet" }""";
        var files = new FakeFiles().With(Settings, theirs);

        var outcome = new ClaudeSettingsEditor(files, Windows).Remove(Folder);

        Assert.False(outcome.Changed);
        Assert.Equal(theirs, files.Read(Settings));
        Assert.Empty(files.Writes);
    }

    [Fact]
    public void Removal_from_an_empty_folder_changes_nothing()
    {
        var files = new FakeFiles();

        var outcome = new ClaudeSettingsEditor(files, Windows).Remove(Folder);

        Assert.False(outcome.Changed);
        Assert.False(files.Has(Settings));
    }

    [Fact]
    public void A_file_Aiko_may_not_touch_is_reported_and_not_lost()
    {
        var files = new FakeFiles().With(Settings, """{ "model": "opus" }""");
        files.Unreadable.Add(Settings);

        var outcome = new ClaudeSettingsEditor(files, Windows).Add(Folder, Command());

        Assert.False(outcome.Changed);
        Assert.NotEqual(PatchProblem.None, outcome.Problem);
        Assert.Empty(files.Writes);
    }

    [Fact]
    public void Without_a_bridge_path_nothing_is_written()
    {
        var files = new FakeFiles();

        var outcome = new ClaudeSettingsEditor(files, Windows).Add(Folder, string.Empty);

        Assert.False(outcome.Changed);
        Assert.NotEqual(PatchProblem.None, outcome.Problem);
        Assert.Empty(files.Writes);
    }

    [Fact]
    public void Repair_fixes_our_line_after_a_reinstall_elsewhere()
    {
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());

        var outcome = editor.RepairIfOurs(Folder, Command(Moved));

        Assert.True(outcome.Changed);
        Assert.Contains(Moved.Replace(@"\", @"\\"), files.Read(Settings));
    }

    [Fact]
    public void Repair_never_installs_the_line_for_somebody_who_said_no()
    {
        // The wizard's "not now" is an answer, and startup is not the place to overrule it.
        var files = new FakeFiles().With(Settings, """{ "model": "opus" }""");

        var outcome = new ClaudeSettingsEditor(files, Windows).RepairIfOurs(Folder, Command());

        Assert.False(outcome.Changed);
        Assert.Empty(files.Writes);
        Assert.DoesNotContain("statusLine", files.Read(Settings));
    }

    [Fact]
    public void Repair_leaves_somebody_elses_line_alone()
    {
        var theirs = """{ "statusLine": { "type": "command", "command": "mine.sh" } }""";
        var files = new FakeFiles().With(Settings, theirs);

        var outcome = new ClaudeSettingsEditor(files, Windows).RepairIfOurs(Folder, Command());

        Assert.False(outcome.Changed);
        Assert.Equal(theirs, files.Read(Settings));
    }

    [Fact]
    public void Repair_with_nothing_to_fix_writes_nothing()
    {
        var files = new FakeFiles();
        var editor = new ClaudeSettingsEditor(files, Windows);
        editor.Add(Folder, Command());
        files.Writes.Clear();

        var outcome = editor.RepairIfOurs(Folder, Command());

        Assert.False(outcome.Changed);
        Assert.Empty(files.Writes);
    }
}

File: mac/Tests/AikoKitTests/ClaudeSettingsEditorTests.swift
import Foundation
import Testing

@testable import AikoKit

/// Writing to a file that belongs to another program.
///
/// Removal walks every Claude Code folder on the machine and rewrites what it finds, and until now
/// none of it had a test. These cover the paths that end badly: no backup, a backup already there,
/// a file the user changed after installing, and a folder Aiko cannot read.
struct ClaudeSettingsEditorTests {
    static let folder = #"C:\Users\someone\.claude"#
    static let bridge = #"C:\Users\someone\AppData\Local\Aiko\current\Aiko.Bridge.exe"#
    static let moved = #"D:\Aiko\Aiko.Bridge.exe"#

    static var settings: String { ClaudeSettingsEditor.pathIn(folder) }
    static var backup: String { ClaudeSettingsEditor.backupPathIn(folder) }

    private func command(_ path: String = ClaudeSettingsEditorTests.bridge) -> String {
        BridgeCommand.forPath(path)
    }

    private func editor(_ files: FakeFiles) -> ClaudeSettingsEditor {
        ClaudeSettingsEditor(files: files, patch: SettingsJsonPatch(.windows))
    }

    @Test
    func aCopyIsMadeBeforeTheFirstChange() {
        let original = #"{ "statusLine": { "type": "command", "command": "mine.sh" } }"#
        let files = FakeFiles().with(Self.settings, original)

        _ = editor(files).add(Self.folder, command())

        #expect(files.has(Self.backup))
        #expect(files.read(Self.backup) == original)
    }

    @Test
    func aCopyThatAlreadyExistsIsLeftAlone() {
        // A second copy would save our own edit and lose what the user had.
        let files = FakeFiles()
            .with(Self.settings, #"{ "statusLine": { "type": "command", "command": "new.sh" } }"#)
            .with(Self.backup, #"{ "statusLine": { "type": "command", "command": "original.sh" } }"#)

        _ = editor(files).add(Self.folder, command())

        #expect(files.read(Self.backup).contains("original.sh"))
    }

    @Test
    func aFolderWithNoSettingsFileGetsOne() {
        let files = FakeFiles()

        let outcome = editor(files).add(Self.folder, command())

        #expect(outcome.changed)
        #expect(files.has(Self.settings))
        // Nothing existed, so there was nothing worth copying.
        #expect(!files.has(Self.backup))
    }

    @Test
    func theFileIsWrittenThroughATemporaryOne() {
        // A half written settings file breaks Claude Code, not just Aiko.
        let files = FakeFiles()

        _ = editor(files).add(Self.folder, command())

        #expect(files.writes.contains { $0.hasSuffix(".aiko.tmp") })
        #expect(!files.writes.contains { $0 == Self.settings })
    }

    @Test
    func addingOurLineTwiceWritesNothingTheSecondTime() {
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())
        files.writes.removeAll()

        let outcome = editor.add(Self.folder, command())

        #expect(!outcome.changed)
        #expect(files.writes.isEmpty)
    }

    @Test
    func removalPutsBackTheLineTheUserHad() {
        let files = FakeFiles().with(Self.settings, #"{ "statusLine": { "type": "command", "command": "mine.sh" } }"#)
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        let outcome = editor.remove(Self.folder)

        #expect(outcome.changed)
        #expect(files.read(Self.settings).contains("mine.sh"))
        #expect(!files.read(Self.settings).contains(SettingsJsonPatch.wrappedKey))
    }

    @Test
    func removalDropsTheCopyItMade() {
        let files = FakeFiles().with(Self.settings, #"{ "statusLine": { "type": "command", "command": "mine.sh" } }"#)
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        _ = editor.remove(Self.folder)

        // The copy was insurance while Aiko was in the file, and the readme promises there is
        // nothing left to clean up by hand.
        #expect(!files.has(Self.backup))
    }

    @Test
    func removalKeepsWhatTheUserChangedAfterInstalling() throws {
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        // They edited the file themselves while Aiko was in it.
        var edited = try #require(JsonNode.parse(files.read(Self.settings))?.objectValue)
        edited["model"] = .string("opus")
        files.with(Self.settings, JsonNode.object(edited).toJsonString())

        _ = editor.remove(Self.folder)

        // Restoring the copy wholesale would have thrown this away.
        #expect(files.read(Self.settings).contains("opus"))
    }

    @Test
    func removalDeletesASettingsFileAikoMadeFromNothing() {
        // Windows Sandbox: no settings.json before Aiko, and "{}" left behind after it.
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        _ = editor.remove(Self.folder)

        #expect(!files.has(Self.settings))
    }

    @Test
    func removalKeepsAFileAikoMadeOnceSomethingElseWasWrittenToIt() throws {
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        var edited = try #require(JsonNode.parse(files.read(Self.settings))?.objectValue)
        edited["model"] = .string("opus")
        files.with(Self.settings, JsonNode.object(edited).toJsonString())

        _ = editor.remove(Self.folder)

        #expect(files.has(Self.settings))
        #expect(files.read(Self.settings).contains("opus"))
    }

    @Test
    func removalKeepsAnEmptyFileThatWasThereBeforeAiko() {
        // It was the user's file, even if it said nothing. A copy was made of it, and that copy is
        // what tells the two cases apart.
        let files = FakeFiles().with(Self.settings, "{}")
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        _ = editor.remove(Self.folder)

        #expect(files.has(Self.settings))
    }

    @Test
    func removalFromAFolderAikoNeverTouchedChangesNothing() {
        let theirs = #"{ "model": "sonnet" }"#
        let files = FakeFiles().with(Self.settings, theirs)

        let outcome = editor(files).remove(Self.folder)

        #expect(!outcome.changed)
        #expect(files.read(Self.settings) == theirs)
        #expect(files.writes.isEmpty)
    }

    @Test
    func removalFromAnEmptyFolderChangesNothing() {
        let files = FakeFiles()

        let outcome = editor(files).remove(Self.folder)

        #expect(!outcome.changed)
        #expect(!files.has(Self.settings))
    }

    @Test
    func aFileAikoMayNotTouchIsReportedAndNotLost() {
        let files = FakeFiles().with(Self.settings, #"{ "model": "opus" }"#)
        files.unreadable.insert(Self.settings)

        let outcome = editor(files).add(Self.folder, command())

        #expect(!outcome.changed)
        #expect(outcome.problem != .none)
        #expect(files.writes.isEmpty)
    }

    @Test
    func withoutABridgePathNothingIsWritten() {
        let files = FakeFiles()

        let outcome = editor(files).add(Self.folder, "")

        #expect(!outcome.changed)
        #expect(outcome.problem != .none)
        #expect(files.writes.isEmpty)
    }

    @Test
    func repairFixesOurLineAfterAReinstallElsewhere() {
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())

        let outcome = editor.repairIfOurs(Self.folder, command(Self.moved))

        #expect(outcome.changed)
        #expect(files.read(Self.settings).contains(Self.moved.replacingOccurrences(of: #"\"#, with: #"\\"#)))
    }

    @Test
    func repairNeverInstallsTheLineForSomebodyWhoSaidNo() {
        // The wizard's "not now" is an answer, and startup is not the place to overrule it.
        let files = FakeFiles().with(Self.settings, #"{ "model": "opus" }"#)

        let outcome = editor(files).repairIfOurs(Self.folder, command())

        #expect(!outcome.changed)
        #expect(files.writes.isEmpty)
        #expect(!files.read(Self.settings).contains("statusLine"))
    }

    @Test
    func repairLeavesSomebodyElsesLineAlone() {
        let theirs = #"{ "statusLine": { "type": "command", "command": "mine.sh" } }"#
        let files = FakeFiles().with(Self.settings, theirs)

        let outcome = editor(files).repairIfOurs(Self.folder, command())

        #expect(!outcome.changed)
        #expect(files.read(Self.settings) == theirs)
    }

    @Test
    func repairWithNothingToFixWritesNothing() {
        let files = FakeFiles()
        let editor = editor(files)
        _ = editor.add(Self.folder, command())
        files.writes.removeAll()

        let outcome = editor.repairIfOurs(Self.folder, command())

        #expect(!outcome.changed)
        #expect(files.writes.isEmpty)
    }
}
